src/components/application_manager/rpc_plugins/rc_rpc_plugin/src/commands/mobile/get_interior_vehicle_data_request.cc

```diff
@@ -93,26 +90,114 @@ void GetInteriorVehicleDataRequest::Execute() {
+
+  const auto& request_msg_params = (*message_)[app_mngr::strings::msg_params];
+  LOG4CXX_DEBUG(logger_,
+                "kSubscribe exist" << request_msg_params.keyExists(
+                    message_params::kSubscribe));
+  if (request_msg_params.keyExists(message_params::kSubscribe)) {
+    response_msg_params[message_params::kIsSubscribed] =
+        request_msg_params[message_params::kSubscribe].asBool();
+    if (request_msg_params[message_params::kSubscribe].asBool()) {
+      auto extension = RCHelpers::GetRCExtension(*app);
+      DCHECK(extension);
+      extension->SubscribeToInteriorVehicleData(ModuleType());
+    }
+  }
+  SendResponse(
+      true, mobile_apis::Result::SUCCESS, nullptr, &response_msg_params);
+  if (AppShouldBeUnsubscribed()) {
+    auto extension = RCHelpers::GetRCExtension(*app);
+    DCHECK(extension);
+    extension->UnsubscribeFromInteriorVehicleData(ModuleType());
+  }
+}
+
+bool GetInteriorVehicleDataRequest::CheckRateLimits() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const uint32_t current_requests_amount =
+      interior_data_cache_.GetCurrentAmountOfRequests(ModuleType());
+  LOG4CXX_DEBUG(logger_,
+                "Current amount of requests in the same time frame is: "
+                    << current_requests_amount);
+  if (current_requests_amount < max_request_in_time_frame_) {
+    interior_data_cache_.IncrementAmountOfRequests(ModuleType());
+    return true;
+  }
+  return false;
+}
+
+bool GetInteriorVehicleDataRequest::AppShouldBeUnsubscribed() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const auto& msg_params = (*message_)[app_mngr::strings::msg_params];
+  if (msg_params.keyExists(message_params::kSubscribe)) {
+    return !(msg_params[message_params::kSubscribe].asBool());
+  }
+  return false;
+}
+
+bool GetInteriorVehicleDataRequest::TheLastAppShouldBeUnsubscribed(
+    app_mngr::ApplicationSharedPtr app) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  if (AppShouldBeUnsubscribed()) {
+    const auto subscribed_to_module_type =
+        AppsSubscribedToModuleType(ModuleType());
+    if (subscribed_to_module_type.size() == 1 &&
+        subscribed_to_module_type.front() == app) {
+      LOG4CXX_DEBUG(logger_,
+                    "The last application unsubscribes from " << ModuleType());
+      return true;
+    }
+  }
+  return false;
+}
+
+void GetInteriorVehicleDataRequest::Execute() {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  if (!ProcessCapabilities()) {
     return;
   }
 
   app_mngr::ApplicationSharedPtr app =
       application_manager_.application(connection_key());
 
-  if (HasRequestExcessiveSubscription()) {
-    excessive_subscription_occured_ = true;
-    is_subscribed =
-        (*message_)[app_mngr::strings::msg_params][message_params::kSubscribe]
-            .asBool();
-    RemoveExcessiveSubscription();
+  if (TheLastAppShouldBeUnsubscribed(app) ||
+      !interior_data_cache_.Contains(ModuleType())) {
+    if (HasRequestExcessiveSubscription()) {
```

Comment:
@LuxoftAKutsan Need to check if we need this check

Reply:
@LitvinenkoIra @ZhdanovP we need this check is we are trying to unsubscribe application that is already unsubscribed. IN that case, SDL should remove `subscribe=false` from HMI request